griptape/engines/eval/eval_engine.py

```diff
@@ -0,0 +1,144 @@
+    name: str = field(
+        default=Factory(lambda self: self.id, takes_self=True),
+        metadata={"serializable": True},
+    )
+    criteria: Optional[str] = field(default=None, metadata={"serializable": True})
+    evaluation_steps: Optional[list[str]] = field(default=None, metadata={"serializable": True})
+    prompt_driver: BasePromptDriver = field(default=Factory(lambda: Defaults.drivers_config.prompt_driver))
+    evaluation_params: list[EvalEngine.Param] = field(
+        default=Factory(lambda: [EvalEngine.Param.INPUT, EvalEngine.Param.ACTUAL_OUTPUT])
+    )
+    generate_steps_system_template: J2 = field(default=Factory(lambda: J2("engines/eval/steps/system.j2")))
+    generate_steps_user_template: J2 = field(default=Factory(lambda: J2("engines/eval/steps/user.j2")))
+    generate_results_system_template: J2 = field(default=Factory(lambda: J2("engines/eval/results/system.j2")))
+    generate_results_user_template: J2 = field(default=Factory(lambda: J2("engines/eval/results/user.j2")))
+
+    @criteria.validator  # pyright: ignore[reportAttributeAccessIssue, reportOptionalMemberAccess]
+    def validate_criteria(self, _: Attribute, value: Optional[str]) -> None:
+        if value is None:
+            if self.evaluation_steps is None:
+                raise ValueError("either criteria or evaluation_steps must be specified")
+            return
+
+        if self.evaluation_steps is not None:
+            raise ValueError("can't have both criteria and evaluation_steps specified")
+
+        if not value:
+            raise ValueError("criteria must not be empty")
+
+    @evaluation_steps.validator  # pyright: ignore[reportAttributeAccessIssue, reportOptionalMemberAccess]
+    def validate_evaluation_steps(self, _: Attribute, value: Optional[list[str]]) -> None:
+        if value is None:
+            if self.criteria is None:
+                raise ValueError("either evaluation_steps or criteria must be specified")
+            return
+
+        if self.criteria is not None:
+            raise ValueError("can't have both evaluation_steps and criteria specified")
+
+        if not value:
+            raise ValueError("evaluation_steps must not be empty")
+
+    @evaluation_params.validator  # pyright: ignore[reportAttributeAccessIssue]
+    def validate_evaluation_params(self, attribute: Attribute, value: list[EvalEngine.Param]) -> None:
+        if not value:
+            raise ValueError("evaluation_params must not be empty")
+
+        if EvalEngine.Param.INPUT not in value:
+            raise ValueError("Input is required in evaluation_params")
+
+        if EvalEngine.Param.ACTUAL_OUTPUT not in value:
+            raise ValueError("Actual Output is required in evaluation_params")
+
+    def __attrs_post_init__(self) -> None:
+        if self.evaluation_steps is None:
+            with validators.disabled():
+                self.evaluation_steps = self._generate_steps()
```

Comment:
We're invoking an LLM when constructing the eval engine? Not sure how I feel about this.

Reply:
agreed, this seems like it can be lazily instantiated via a `property` or just directly called during `_generate_results`